Adds the ability to invoke tests with command line arguments

chick opened the pull request:
The major case for this is that command line arguments have a better discovery mechanism through `--help`

- Add ChiseTestShell that can parse command line arguments and produce corresponding annotations
- Adds helper method `withFlags` that is the equivalent for CLI flags as `withAnnotations`
- Slight refactoring of TestBuilder to be a trait with two implementations AnnotatedTestBuilder and FlaggedTestBuilder
- Changed experimental basic verilator test to use this new way of selecting the verilator backend

seldridge commented:
High level: I'm in favor of a unified test builder that takes both options and annotations. Since command line options are convertible to annotations, it's reasonable to have all these together.

Stages work this way with their `def execute(options: Array[String], annotations: AnnotationSeq)` method.

chick commented:
@seldridge The user facing API here looks like
`test(dutGen).withAnnotations(annos)` or `test(dutGen).withFlags(flags)`. I guess this would mean that for completeness there should be a
`test(dutGen).withFlagsAndAnnotations(flags, annos)`

chick commented:
Simplified TestBuilder to minimize code duplication and class proliferation
Add new methods .withFlags and .withAnnotationsAndFlags to test() launcher
Moved and simplified tests of new code into OptionsPassingTests

ducky64 commented:
Um, withAnnotationsAndFlags seems a bit excessive... and unnecessary since withFlags and withAnnos should compose? It seems like the reason Stages has the combined API is that because there's no composition mechanism, but that isn't the case here?

seldridge commented:
Yeah, to clarify, I'm in favor of the compositional API of:

```scala
test
  .withFlags(Array("--foo", "--bar"))
  .withAnnotations(Seq(BazAnnotation(Qux)))
```

chick commented:
Removed `.withAnnotationsAndFlags`. `.withAnnotations` and `.withFlags` compose properly already.

ducky64 commented on src/test/scala/chiseltest/tests/OptionsPassingTest.scala:
```diff
@@ -6,10 +6,12 @@ import chisel3._
 import treadle.{VerboseAnnotation, WriteVcdAnnotation}
 
-class OptionsPassingTest extends FlatSpec with ChiselScalatestTester with Matchers {
+class OptionsPassingTest extends FlatSpec with ChiselScalatestTester with Matchers with LazyLogging {
```
Whats this LazyLogging for?

chick replied:
Leftover from an earlier test attempt. Removed

ducky64 commented on src/main/scala/chiseltest/experimental/ChiselTestShell.scala:
```diff
@@ -0,0 +1,36 @@
+import firrtl.options.Shell
+import firrtl.stage.FirrtlCli
+
+trait ChiselTestCli extends ChiselCli with FirrtlCli {
```
What is this and how is this used? Should there be example code in the scaladoc?

chick replied:
This only used internally right now to parse the command line options strings in to a Seq of annotations. It attaches the ChiselTest specific options to options available from the upstream stack (chisel, treadle, firrtl)

ducky64 replied:
ok, I see it now

chick commented on src/main/scala/chiseltest/ChiselScalatestTester.scala:
```diff
@@ -3,30 +3,43 @@
+    /** Add a targetDir if none has been specified, check the context
+      * to see if a writeVcd should be added
+      */
+    def updateAnnotations(annotationSeq: AnnotationSeq): AnnotationSeq = {
```
Inlined the function since it was only used once

ucbjrl commented:
The test `allow specifying configuration options using annotations and CLI style flags` is failing with current master versions of firrtl and chisel3. I thought it might have something to do with freechipsproject/firrtl#1392, but undoing that change doesn't appear to fix it.

chick commented:
I can't reproduce this error locally (with latest chisel3, firrtl). what is the error?

ucbjrl commented:
```
[info] - should allow specifying configuration options using annotations and CLI style flags *** FAILED ***
Total FIRRTL Compile Time: 78.2 ms
[info]   false was not true (OptionsPassingTest.scala:65)
 ...
[info] Tests: succeeded 101, failed 1, canceled 0, ignored 2, pending 0
[info] *** 1 TEST FAILED ***
[info] utest
[info] Tests: 1, Passed: 1, Failed: 0
[error] Failed: Total 103, Failed 1, Errors 0, Passed 102, Ignored 2
[error] Failed tests:
[error] 	chiseltest.tests.OptionsPassingTest
[error] (Test / test) sbt.TestsFailedException: Tests unsuccessful
[error] Total time: 141 s (02:21), completed Feb 21, 2020, 12:32:30 AM
```
